compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -2353,9 +2353,13 @@ class Typer extends Namer
           !tree.symbol.isConstructor &&
           !tree.symbol.is(TransparentMethod) &&
           !ctx.mode.is(Mode.Pattern) &&
-          !(isSyntheticApply(tree) && !isExpandableApply))
+          !(isSyntheticApply(tree) && !isExpandableApply)) {
+        if (!pt.classSymbol.hasAnnotation(defn.FunctionalInterfaceAnnot) &&
+            !defn.isFunctionType(pt) &&
+            SAMType.unapply(pt).isDefined)
+          ctx.warning(ex"${tree.symbol} is eta-expanded even though ${pt.classSymbol} does not have the @FunctionalInterface annotation.", tree.pos)
```

Comment:
Small nitpick: I'd change the condition to

    if (!defn.isFunctionType(pt))
      pt match {
        case SAMType(_) if !pt.classSymbol.hasAnnotation(defn.FunctionalInterfaceAnnot) =>
          ...
        case _ =>
     }